doc/dev/dialects.rst

```diff
@@ -0,0 +1,417 @@
+
+
+Introduction to MLIR dialects
+=============================
+
+A *dialect* in MLIR is intended to represent a certain abstraction level, computing domain, or
+set of thematically-connected program elements, by grouping all necessarily IR objects into
+a common namespace. In MLIR, the relevant IR objects include operations, types, attributes, traits,
+interfaces, and passes (among others).
+
+A well-scoped dialect can form a reusable component across different compiler IRs, for example the
+built-in ``func`` dialect provides everything needed to represent functions (including function
+definitions & declarations), call graphs, and inlining transformations.
+This means dialects are not monolithic program representations, but instead can be composed with
+other dialects to represent a large variety of programs. This structure is especially well-suited to
+represent heterogenous programs (like quantum algorithms).
+
+Additionally, dialects are typically organized in an *abstraction hierarchy*, with conversion rules
+describing how programs can be lowered from a high-level representation (or abstraction) to a
+low-level one. One such conversion rule might involve the decomposition of complex instructions
+into a sequence of simpler instructions, or the allocation of memory for various high-level objects.
+Alongside of optimization rules, these processes form the core part of program *compilation*.
+
+For additional information on MLIR dialects, and the type of IR objects provided by MLIR, refer to
+the following documentation pages:
+
+- `the MLIR language reference <https://mlir.llvm.org/docs/LangRef/>`_: A technical reference for
+  the MLIR language (the textual IR format), as well as its concepts and building blocks.
+- `built-in MLIR dialects <https://mlir.llvm.org/docs/Dialects/>`_: A list of dialects and
+  associated documentation for dialects developed with the MLIR project or contributed upstream.
+- `defining MLIR dialects <https://mlir.llvm.org/docs/DefiningDialects/>`_: Documentation describing
+  how to define MLIR dialects and all its components.
+- `debugging tips <https://mlir.llvm.org/getting_started/Debugging/>`_: Useful reference for any
+  kind of MLIR development, especially the debugging flags for `mlir-opt` like tools.
+
+The remainder of this document will guide you through, in concrete terms, how to add a new dialect
+to Catalyst.
+
+Catalyst dialect structure
+==========================
+
+A strong design principle in the MLIR project is the avoidance of boilerplate and redundant code,
+favoring a single source of truth to specify IR components. This is achieved via one or more
+descriptive languages based on the `TableGen <https://llvm.org/docs/TableGen/index.html>`_ format
+known from LLVM. Consequently, dialects, as well as many operations, types, attributes, and other
+IR objects, are specified in two parts:
+
+- a *declarative* part: The main way to define new IR objects via a simple language called
```

Comment:
Minor/optional: bolding the text will help it stand out more, making it easier to parse/skim read
```suggestion
- **a declarative part**: The main way to define new IR objects via a simple language called
```